src/index/ivf/ivf.cc

```diff
@@ -219,77 +209,54 @@ IvfIndexNode<T>::Train(const DataSet& dataset, const Config& cfg) {
             qzr = new (std::nothrow) typename QuantizerT<T>::type(dim, metric.value());
-            index = new (std::nothrow) T(qzr, dim, MatchNlist(dataset.GetRows(), ivf_flat_cfg.nlist), metric.value());
+            index_ = std::make_shared<faiss::IndexIVFFlat>(qzr, dim, nlist, metric.value());
+            index_->own_fields = true;
+            index_->train(rows, (const float*)data);
         }
         if constexpr (std::is_same<faiss::IndexIVFPQ, T>::value) {
             const IvfPqConfig& ivf_pq_cfg = static_cast<const IvfPqConfig&>(cfg);
+            auto nlist = MatchNlist(rows, ivf_pq_cfg.nlist);
+            auto nbits = MatchNbits(rows, ivf_pq_cfg.nbits);
             qzr = new (std::nothrow) typename QuantizerT<T>::type(dim, metric.value());
-            index = new (std::nothrow) T(qzr, dim, MatchNlist(dataset.GetRows(), ivf_pq_cfg.nlist), ivf_pq_cfg.m,
-                                         MatchNbits(dataset.GetRows(), ivf_pq_cfg.nbits), metric.value());
+            index_ = std::make_shared<faiss::IndexIVFPQ>(qzr, dim, nlist, ivf_pq_cfg.m, nbits, metric.value());
+            index_->own_fields = true;
+            index_->train(rows, (const float*)data);
         }
         if constexpr (std::is_same<faiss::IndexIVFScalarQuantizer, T>::value) {
             const IvfSqConfig& ivf_sq_cfg = static_cast<const IvfSqConfig&>(cfg);
+            auto nlist = MatchNlist(rows, ivf_sq_cfg.nlist);
             qzr = new (std::nothrow) typename QuantizerT<T>::type(dim, metric.value());
-            index = new (std::nothrow) T(qzr, dim, MatchNlist(dataset.GetRows(), ivf_sq_cfg.nlist),
-                                         faiss::QuantizerType::QT_8bit, metric.value());
+            index_ = std::make_shared<faiss::IndexIVFScalarQuantizer>(qzr, dim, nlist, faiss::QuantizerType::QT_8bit,
+                                                                      metric.value());
+            index_->own_fields = true;
+            index_->train(rows, (const float*)data);
         }
         if constexpr (std::is_same<faiss::IndexBinaryIVF, T>::value) {
             const IvfBinConfig& ivf_bin_cfg = static_cast<const IvfBinConfig&>(cfg);
+            auto nlist = MatchNlist(rows, ivf_bin_cfg.nlist);
             qzr = new (std::nothrow) typename QuantizerT<T>::type(dim, metric.value());
-            index = new (std::nothrow) T(qzr, dim, MatchNlist(dataset.GetRows(), ivf_bin_cfg.nlist), metric.value());
-        }
-        if (qzr == nullptr || index == nullptr) {
-            if (qzr) {
-                delete qzr;
-            }
-            if (index) {
-                delete index;
-            }
-            LOG_KNOWHERE_WARNING_ << "memory malloc error.";
-            return Status::malloc_error;
+            index_ = std::make_shared<faiss::IndexBinaryIVF>(qzr, dim, nlist, metric.value());
+            index_->own_fields = true;
+            index_->train(rows, (const uint8_t*)data);
```

Comment:
if this place index_ make_shared failed, or train failed, the old index hold on  var index_ maybe breaked.

Reply:
one segment will only be trained once, "index_" must be nullptr before Train